### DISCUSSION TITLE
Helmholtz Property Package for R32 Compression does not intitialise

### DISCUSSION THREAD

**Discussion by bjl25** [IDAES/idaes-pse#1271] [Q&A]
I have started to model an R32 heat pump system in IDAES and have selected the HelmholtzParameterBlock to model the properties of R32. 

I started by modeling the compressor to test how the package performs. I had the following major code snippets:

m.fs.propertiesRC = HelmholtzParameterBlock(
  pure_component="R32",
  phase_presentation=PhaseType.MIX,
  state_vars=StateVars.PH
)

m.fs.Comp = PressureChanger(
    property_package=m.fs.propertiesRC,
    compressor=True,
    thermodynamic_assumption=ThermodynamicAssumption.isentropic,
)

h1= m.fs.propertiesRC.htpx(288 * units.K, 813100 * units.Pa)
print(h1)
m.fs.Comp.inlet.flow_mol.fix(50)
#m.fs.Comp.inlet.temperature[0].fix(288)
m.fs.Comp.inlet.enth_mol.fix(h1)
m.fs.Comp.inlet.pressure.fix(813100)

m.fs.Comp.outlet.pressure.fix(3248913)
m.fs.Comp.efficiency_isentropic.fix(0.70)

DOF_final = degrees_of_freedom(m)

I received a warning stating my molar enthalpy was outside the bounds then flags = blk.initialize_build(*args, **kwargs) portion failed to initialize. I checked the enthlpy generated from Helmholtz against Coolprop and it matched. 

I then changed the phase-type to vapor only and the StateVars to TPx which subsequently initialized fine and produced the same answer as my Excel model. I then changed the state variables back to PH, with the phase type still set to Vapour only, and produced the same error. Finally, I tested a mixed phase type with TPx as the state variable and initialized fine but got dramatically reduced work input. 

Please help I would love to use this program further for modeling heat pumps.
PS Code is below and sorry for the mess I was testing out connecting the unit ops together and creating a R32 PR property package.

[Heat Pump.txt](https://github.com/IDAES/idaes-pse/files/12680004/Heat.Pump.txt)

**Comment by andrewlee94** [comment]
@bjl25 Could you post the full error traceback and warnings you are seeing? It is hard to tell exactly what is going on from just your description.

However the fact that you are getting a warning about a value outside of bounds is a strong indication that the values you are using are not feasible. Also, note that you have to be careful when directly compare values between different implementations of properties (i.e. checking values against CoolProp). You have to make sure the reference states are the same, or apply an offset to align the values (although it is likely that the reference state is the same here).

**Comment by bjl25** [comment]
Thank you for the reply @andrewlee94 the full error traceback is below. To re-explain basically the solver converges when I use R32 Helmholtz in vapor only and TPx mode. If I use Pressure Enthalpy mode for the property package it fails to converge and if I use the LG or mixed for the phase type with TPx receive a different answer. (The pressure and enthalpy I use is calculated using htpx with the same T and P from the case that converges). 


The traceback error for PH StateVars mode:
Enthalpy is 27704.73042940659
WARNING (W1002): Setting Var
'fs.Comp.control_volume.properties_in[0.0].enth_mol' to a numeric value
`27704.73042940659` outside the bounds (-992.1972213327272,
26013.433049962445).
    See also https://pyomo.readthedocs.io/en/stable/errors.html#w1002
The DOF is 0
WARNING (W1002): Setting Var
'fs.Comp.control_volume.properties_out[0.0].enth_mol' to a numeric value
`27704.73042940659` outside the bounds (-992.1972213327272,
26013.433049962445).
    See also https://pyomo.readthedocs.io/en/stable/errors.html#w1002
WARNING (W1002): Setting Var 'fs.Comp.properties_isentropic[0.0].enth_mol' to
a numeric value `27704.73042940659` outside the bounds (-992.1972213327272,
26013.433049962445).
    See also https://pyomo.readthedocs.io/en/stable/errors.html#w1002
Traceback (most recent call last):
  File "c:\Users\bjl25\Documents\GitHub\Bens-IDAES-Ideas\IDAES\Models\Bens-IDAES-Ideas\IDAES\Models\Test cases Working\HeatPump.py", line 127, in <module>
    m.fs.Comp.initialize(outlvl=idaeslog.INFO)
  File "C:\Users\bjl25\AppData\Local\anaconda3\envs\IdaesBen\Lib\site-packages\idaes\core\base\unit_model.py", line 540, in initialize
    flags = blk.initialize_build(*args, **kwargs)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\bjl25\AppData\Local\anaconda3\envs\IdaesBen\Lib\site-packages\idaes\models\unit_models\pressure_changer.py", line 876, in initialize_build     
    blk.init_isentropic(
  File "C:\Users\bjl25\AppData\Local\anaconda3\envs\IdaesBen\Lib\site-packages\idaes\models\unit_models\pressure_changer.py", line 1181, in init_isentropic     
    raise InitializationError(
idaes.core.util.exceptions.InitializationError: fs.Comp failed to initialize successfully. Please check the output logs for more information.

**Comment by bjl25** [comment]
@andrewlee94 Here are the results for the same conditions 290K, 8.1 bar to 32.4 bar inlet vapor fraction of 1 (for the mix phase case) just changing the PhaseType. 

Gas Only:
EXIT: Optimal Solution Found.
outlet : Size=1
    Key  : Name        : Value
    None :    flow_mol : {0.0: 50.0}
         :    pressure : {0.0: 3248913}
         : temperature : {0.0: 401.72278282298157}
work_mechanical : Size=1, Index=fs._time, ReferenceTo=fs.Comp.control_volume.work[:]
    Key : Lower : Value              : Upper : Fixed : Stale : Domain
    0.0 :  None : 232658.69671592495 :  None : False : False :  Reals

Mix:
    Key  : Name        : Value
    None :    flow_mol : {0.0: 50.0}
         :    pressure : {0.0: 3248913}
         : temperature : {0.0: 396.6361424653954}
         :  vapor_frac : {0.0: 0.9999999999999998}
work_mechanical : Size=1, Index=fs._time, ReferenceTo=fs.Comp.control_volume.work[:]
    Key : Lower : Value              : Upper : Fixed : Stale : Domain
    0.0 :  None : 210609.40327404859 :  None : False : False :  Reals

**Comment by andrewlee94** [comment]
@bjl25 Firstly, you cannot put any value on a result when the solver returns infeasible or fails to initialize - the solver failed to find a solution so whatever values are there are just what was left when the solver gave up. So, there is no point in comparing the mechanical work between a case which solved and one which didn't.

Secondly, the fact that you are getting warnings about variables outside of bounds is suggesting that you are working outside the feasible range for the given property package, and thus it is not unexpected that it might fail to solve. I would suggest trying to solve the model within the feasible range first to see if that solves - if you see a failure with a feasible value then there is an issue with the model/initialization but if the issue is just the bounds then you need to ask if the range you are solving for is valid.

**Reply by bjl25** [reply]
@andrewlee94 Apologies for not being clear but the values I posted were for cases when no error was thrown (when using the TPX version no initialization error or solution error is thrown, it just gives different results depending on the phase type selected in the setup). The errors only come when using R32 Helmholtz in PH mode but they are the same conditions as TPX mode which do solve with no errors. I was interested in why for the same apparent conditions I can solve and not solve. Is it me misunderstanding the htpx function 

You are correct that if I lower the Pressure in the function h1= m.fs.propertiesRC.htpx(p=81310 * units.Pa, x = 1, amount_basis=AmountBasis.MASS, with_units = True) I can calculate the answer, or though the HP I am modeling operates in those outer regions.

**Comment by andrewlee94** [comment]
The key issue here is this output you showed:

WARNING (W1002): Setting Var
'fs.Comp.control_volume.properties_in[0.0].enth_mol' to a numeric value
27704.73042940659 outside the bounds (-992.1972213327272,
26013.433049962445).

This indicates that the value you are using for enthalpy (from the htpx function) is outside the bounds supported by the property package. The property packages puts upper and lower bounds on the enthalpy to indicate the range over which the correlations are reliable, and it will fail to converge if you use values outside of this . In this case, the enthalpy you are trying to solve for is above the upper bound (27704.73042940659 > 26013.433049962445), thus when you go to solve the model it is infeasible because the bounds are violated.

The issue is not the htpx function, but the conditions you are trying to solve for. You note that you are operating in the outer regions for the refrigerant, and in this case you are pushing further than the model is willing to allow. This doesn’t mean the region is non-physical, but it does indicate that the correlations used were not fitted in this region and thus would require extrapolation and the added uncertainty that implies.

You basically have three options:


  1.  Limit yourself to be within the current bounds of trusted space.
  2.  Find new parameters that cover the space you are interested in and set up a new property package (the Helmholtz property package can accept custom parameters).
  3.  Understanding the risks of extrapolation, relax the bounds set in the property package.

**Reply by bjl25** [reply]
@andrewlee94 Thank you so much for your time and patience in answering all my questions it means a lot. I think I will look into using the modular property package integrated with COOLPROP (which does have the capability for that operating range). Have a good rest of the week.

**Comment by andrewlee94** [comment]
@bjl25 Note that the CoolProp integration is a prototype and only supports cubic equations of state, so it might not actually do what you require. If you want to use Helmholtz equations of state, then you would need to look at loading the parameter values from CoolProp into IDAES. Also note that CoolProp might just be less restrictive on extrapolation than we are but suffer from the same limitations on the fitting data (I believe we both used the same data sources).

**Comment by bjl25** [comment]
@andrewlee94 Hello Andrew, I have circled back around to my heat pump modeling project. You are 100% correct that it is a bounding issue that is causing the initialization issue. I investigated your Helmohotz property package and CoolProps property package and found that they use the same source (https://pubs.aip.org/aip/jpr/article-abstract/26/6/1273/241752/An-International-Standard-Equation-of-State-for?redirectedFrom=fulltext). I found that it is rated for temperatures from the Triple Point at 136.34 K to 435 K and Pressures up to 70 MPa which encompasses my maximum P and T conditions of 400K and 3.2 MPa. 

Therefore I am confident in the property parameters and wish to extend the bounds but I can not figure out how to. Could you please point me in the right direction? 

Thank you

**Reply by andrewlee94** [reply]
The easiest way to do it is to change the `default_[variable]_bounds` in the property parameter block **before** you add any unit models to the flowsheet. The bounds on the state variables are set to a default value as soon as the state block is created so you need to set the default value to use before building the unit models (otherwise you need to go change each bound manually).

E.g.:
```
m.fs.propertiesRC = HelmholtzParameterBlock(
pure_component="R32",
phase_presentation=PhaseType.MIX,
state_vars=StateVars.PH
)

m.fs.propertiesRC.default_pressure_bounds = (Pmin, Pmax)
```

You can also do this for any other state variable in the system (i.e. enthalpy and flow rate) as required.
